streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -253,44 +255,44 @@ trait ZStream[-R, +E, +A] extends Serializable { self =>
-                      e => out.offer(Take.Fail(e)).unit.toManaged_,
-                      // The driver stream had a defect.
-                      c => (out.shutdown *> ZIO.halt(c)).toManaged_
-                    ),
-                    innerFibers =>
+            _ <- self.foreachManaged { a =>
+                  for {
+                    latch <- Promise.make[Nothing, Unit]
+                    innerStream = Stream
+                      .bracket(permits.acquire *> latch.succeed(()))(_ => permits.release)
+                      .flatMap(_ => f(a))
+                      .foreach(b => out.offer(Take.Value(b)).unit)
+                      .foldCauseM(
+                        cause => out.offer(Take.Fail(cause)) *> innerFailure.fail(cause),
+                        _ => ZIO.unit
+                      )
+                    _ <- (innerStream race interruptInners.await).fork
+                    // Make sure that the current inner stream has actually succeeded in acquiring
+                    // a permit before continuing. Otherwise, two bad things happen:
+                    // - we might needlessly fork inner streams without available permits
+                    // - worse, we could reach the end of the stream and acquire the permits ourselves
+                    //   before the inners had a chance to start
+                    _ <- latch.await
+                  } yield ()
+                }.foldCauseM(
+                    cause =>
+                      (interruptInners.succeed(()) *> permits.acquireN(n) *> out
+                        .offer(Take.Fail(cause))).unit.toManaged_,
+                    _ =>
                       innerFailure.await
-                        .raceWith(Fiber.joinAll(innerFibers))(
-                          // One of the inner fibers failed. It already enqueued its failure,
-                          // so we don't need to do anything except interrupt all the other fibers.
-                          leftDone = (_, _) => Fiber.interruptAll(innerFibers),
+                        .raceWith(permits.acquireN(n))(
+                          // One of the inner fibers failed. It already enqueued its failure, so we
+                          // just need to interrupt the inner fibers and acquire all permits to await
+                          // their termination
+                          leftDone = (_, permitAcquisition) => interruptInners.succeed(()) *> permitAcquisition.join,
                           // All fibers completed successfully, so we just need to signal that
                           // we're done.
                           rightDone = (_, failureAwait) => out.offer(Take.End) *> failureAwait.interrupt
                         )
                         .toManaged_
                   )
                   .fork
-            _ <- ZManaged.finalizer(innerFibers.get.flatMap(Fiber.interruptAll))
+            _ <- ZManaged.finalizer(interruptInners.succeed(()))
```

Comment:
Is this still necessary?

Reply:
Yes, I'll add a comment. It handles the case where the stream after the queue is short-circuited.

There's some overlap between the different ending cases (interruption, short circuiting, etc) but I don't want to look at this combinator again this weekend. :-)

Reply:
Actually this isn't quite right. Even though we're short circuiting the driver stream would still try to fork all the inner streams (which would get interrupted immediately). I'll fix this now.

Reply:
Sorry ignore that. `ZManaged#fork` installs a finalizer that interrupts the fiber on the way out so we won't be unnecessarily forking inner streams.